enable async generators as resources

sh-rp opened the pull request:
<!--
Thank you for submitting a pull request! Please provide a brief description of your changes below.
-->
A simple implementation to support async generators in our resources and transformers. This implementation basically converts an async generator to an awaitable list, so it can be run in the futures pool. 

Implementation details: The way it is implemented now, the async generator will block one futures slot in the extraction until it is completed. The question is, should we aim at implementing something that will finish a future regularly after a certain batch size and yield control back to the main thread, but have some kind of info in the FuturePipeItem indicating that there still is an open async generator that it can be rescheduled? Somehow a follow-up future of sorts?

sh-rp commented on dlt/extract/pipe.py:
```diff
@@ -791,6 +806,27 @@ def _get_source_item(self) -> ResolvablePipeItem:
         elif self._next_item_mode == "round_robin":
             return self._get_source_item_round_robin()
 
+    def _get_next_item_from_generator(
```
here there only is refactoring to get rid of non-dry code.

rudolfix commented on dlt/extract/utils.py:
```diff
@@ -119,6 +119,25 @@ def check_compat_transformer(name: str, f: AnyFun, sig: inspect.Signature) -> in
+        result: List[TDataItem] = []
+        try:
+            item: TDataItems = None
+            while item := await f.__anext__():
```
the problem with this thing is that it will consume the whole generator in memory and then return results once. so ie. when there's a transformer it will receive a full dataset. right?

maybe we need to send each f.__next__() to a future pool and wrap this in a function that returns both a next result (after await) and the async iterator itself? so you put it into futures pool again and return the current value

rudolfix reviewed (changes requested):
wrapping is very clever! but we need to yield item by item. we cannot collect all the items and yield at once...

sh-rp commented:
@rudolfix yes you are right, i was already thinking about that after submitting the PR. I have changed it now that the wrapped async generator returns an iterator of awaitables. There is some magic inside which essentially emulates this "yield none" behavior we implemented last summer to make round_robin extract work (I can point you to the commit I think if you don't remember). My remaining questions are:
* Should we maybe change the default item mode to round_robin? I don't think the newbie user will find out about this if they just start yielding stuff
* Should we make the chunk-size of this generator configurable, or are we happy with yielding single items? I suppose we can keep it like this and it will replicate the behavior of the regular generator.
* I have modified the function that evaluates the result of futures a bit to not have stopiteration exceptions be an error, we could also have a custom exception there if you think we should still treat this as an error in other cases. wdyt?

sh-rp commented:
If we follow the approach of adding a follow up future, which i think is doable, then parallelism I think will also work for the fifo mode, I somehow like my approach better though, because it does not modify the pipe code much.

rudolfix commented on dlt/extract/utils.py:
```diff
@@ -119,6 +119,29 @@ def check_compat_transformer(name: str, f: AnyFun, sig: inspect.Signature) -> in
+            is_running = False
+
+    # it is best to use the round robin strategy here if multiple async generators are used in resources
+    while not exhausted:
```
I think we should use the same strategy that user wants? anyway. IMO if you have two async generators they will be executed in parallel in futures, right?

sh-rp replied:
not if we emit a list of awaitables, then one list will be processed first, then the other (you can see that in the test)

rudolfix commented on dlt/extract/utils.py:
```diff
@@ -119,6 +119,29 @@ def check_compat_transformer(name: str, f: AnyFun, sig: inspect.Signature) -> in
+    # it is best to use the round robin strategy here if multiple async generators are used in resources
+    while not exhausted:
+        while is_running:
+            yield None
```
100% cpu usage here? this will be executed in tight loop right? if there's None we go for the next resource and we get None again so we go again etc.

btw. I think we are close and the idea with yielding None is good.

rudolfix commented on dlt/extract/utils.py:
```diff
@@ -119,6 +119,29 @@ def check_compat_transformer(name: str, f: AnyFun, sig: inspect.Signature) -> in
+
+    # it is best to use the round robin strategy here if multiple async generators are used in resources
+    while not exhausted:
+        while is_running:
```
hmmm maybe asyncio critical sections are better? https://docs.python.org/3.8/library/asyncio-sync.html (if they work here because this code is sync)

rudolfix commented:
@sh-rp 

> * Should we maybe change the default item mode to round_robin? I don't think the newbie user will find out about this if they just start yielding stuff
I'd say no. the default behavior is to evaluate resources one by another fully. I think it is good.
we can switch to round robin if there's an async generator in any of the pipes though. otherwise your current implementation will starve the async generators is there are normal ones. btw. there are a lot of nuances due to this hack with yielding None in a loop

> * Should we make the chunk-size of this generator configurable, or are we happy with yielding single items? I suppose we can keep it like this and it will replicate the behavior of the regular generator.
single item can be a list/batch. regular generators work the same way, no?

> * I have modified the function that evaluates the result of futures a bit to not have stopiteration exceptions be an error, we could also have a custom exception there if you think we should still treat this as an error in other cases. wdyt?
I do not see this modification. do you mean this?:
```
if isinstance(ex, StopAsyncIteration):
                return None
```
this is handling this new async iterator and LGTM

but yeah we should test if limiting async iterator still works. could you add a test with does `add_limit` on a resource that is async generator to make sure we close it properly before it is exhausted.

also please add a test where we have async and sync generator to see if we are not starving the async.

sh-rp commented:
@rudolfix do you remember what the reasoning of these lines was?
```
        # if there are currently more sources than added initially, we need to process the new ones first
        if sources_count > self._initial_sources_count:
            return self._get_source_item_fifo()
```
This makes the round robin fall back to fifo if more pipes were added during extraction. I remember that I put this in, but i have no recollection why. I will slow down parallel execution as soon as dynamically creates pipes are present in async generator scenarios and I am wondering if we can remove it.

sultaniman commented on dlt/extract/utils.py:
```diff
@@ -142,8 +178,12 @@ def wrap_resource_gen(
+    ):
 
         def _partial() -> Any:
             # print(f"_PARTIAL: {args} {kwargs}")
```
Is this supposed to be here?

sh-rp commented on dlt/extract/resource.py:
```diff
@@ -310,10 +310,16 @@ def _gen_wrap(gen: TPipeStep) -> TPipeStep:
                 for i in gen:  # type: ignore # TODO: help me fix this later
                     yield i
-                    count += 1
+                    if i is not None:
```
this line is needed for the async generator to work properly, it changes the behavior of the limit, but probably that is ok, i am not sure.

sh-rp commented on dlt/extract/pipe.py:
```diff
@@ -773,91 +792,63 @@ def _resolve_futures(self) -> ResolvablePipeItem:
-        # if there are currently more sources than added initially, we need to process the new ones first
-        if sources_count > self._initial_sources_count:
-            return self._get_source_item_current()
         try:
```
this function now nicely combines fifo and round_robin. in fifo mode it says on the first source and only ventures into the next ones if that returns none. It would be quite easy to switch it back to the old behavior though. I removed this part that switches from round robin to fifo in some cases as it does not really make sense anymore imho if fifo also can switch the source index.

sh-rp replied:
I implemented it this way (and corrected the decreasing of initial sources counter along the way). Now there is a diferentiation between fifo and strict_fifo, I think this is necessary to prevent a scenario as you describe it.

sh-rp commented on dlt/extract/utils.py:
```diff
@@ -119,6 +131,35 @@ def check_compat_transformer(name: str, f: AnyFun, sig: inspect.Signature) -> in
+                yield None
+            yield run()
+    except GeneratorExit:
+        # clean up async generator
```
i don't think I need to clean up here, this can happen at the end of the loop in the pipe

sh-rp replied:
the problem is, that the "parent" generator gets closed before the last item is evaulated that we want to evaluate. so it has to stay this way. do you know what I mean?

sh-rp replied:
But i just had a nice idea of how to close the generator in the right moment this way, which is a good improvement :)

sh-rp commented on tests/pipeline/test_resources_evaluation.py:
```diff
@@ -0,0 +1,280 @@
+
+
+@pytest.mark.parametrize("parallelized", [True, False])
+def test_async_decorator_experiment(parallelized) -> None:
```
this for now is just an experiment, but this could be the extension for the defer decorator. if you put it on a resource function, it will make every iteration run as a future, so quite cool i think.

sh-rp replied:
PS: i know that i was not supposed to do this yet ;)

sh-rp commented on dlt/extract/pipe.py:
```diff
@@ -703,6 +716,12 @@ def stop_background_loop(loop: asyncio.AbstractEventLoop) -> None:
+                self._async_pool.shutdown_asyncgens(), self._ensure_async_pool()
+            )
+            while not future.done():
+                sleep(self.futures_poll_interval)
```
not 100% certain i implemented this right, but i think i did. we need to shut down all the open gens in this pool.

sh-rp replied:
@rudolfix I am not sure what you mean with "async". there is a test called `def test_limit_async_resource() -> None:` which tests limiting async resources (which get turned into generators by the wrapper). Can you maybe explain?

sh-rp replied:
Do you mean async iterators?

sh-rp commented on dlt/extract/utils.py:
```diff
@@ -119,6 +131,38 @@ def check_compat_transformer(name: str, f: AnyFun, sig: inspect.Signature) -> in
+) -> Generator[Awaitable[TDataItems], None, None]:
+    """Wraps an async generator into a list of awaitables"""
+    exhausted = False
+    busy = False
```
using an async.lock here will fail on python 3.8, and 3.9 because there is no current loop on the main thread. i am not sure why it is allowed in python 3.10. threading locks should not be used with coroutines as far as I understand.

rudolfix commented on tests/extract/test_extract_pipe.py:
```diff
@@ -39,13 +39,13 @@ def get_pipes():
-    # items will be round robin, nested iterators are fully iterated and appear inline as soon as they are encountered
-    assert [pi.item for pi in _l] == [1, 11, 20, 2, 12, 21, 55, 56, 77, 88, 89, 13, 3, 14, 4, 15]
+    # items will be round robin, nested iterators are integrated into the round robin
+    assert [pi.item for pi in _l] == [1, 11, 20, 2, 12, 21, 3, 13, 55, 4, 14, 56, 15, 77, 88, 89]
```
please test a case when we have too many nested iterators

sh-rp replied:
maybe a couple more tests would be not so bad here, especially since it is a bit hard to read this test (ie to verify that it is correct) :)

rudolfix reviewed (changes requested):
thx for this! I think we fixed many flaws with parallelism, not only async gens
1. we need a test where exception happens during async generation. I have a set of such tests somewhere, please add this one
2. we need a test where large amount of nested iterators is created, see review
3. we should move defer decorator (we can alias it!) to separate PR. nevertheless I reviewed it :)

sh-rp commented on dlt/extract/pipe.py:
```diff
@@ -613,8 +614,19 @@ def __next__(self) -> PipeItem:
+                continue
+
+            # handle async iterator items as new source
+            if inspect.isasyncgen(item):
```
yeah you are totally right, this is a leftover from something else, I updated it in the other places too.

sh-rp commented on tests/pipeline/test_resources_evaluation.py:
```diff
@@ -0,0 +1,285 @@
+            gen = f(*args, **kwargs)
+            # unpack generator
+            if inspect.isfunction(gen):
+                gen = gen()
```
next pr!

sh-rp commented on tests/pipeline/test_resources_evaluation.py:
```diff
@@ -0,0 +1,285 @@
+            if inspect.isfunction(gen):
+                gen = gen()
+            # if we have an async gen, no further action is needed
+            if inspect.isasyncgen(gen):
```
next pr!

sh-rp commented on tests/pipeline/test_resources_evaluation.py:
```diff
@@ -0,0 +1,285 @@
+    threads = set()
+
+    def parallelize(f) -> Any:
+        """converts regular itarable to generator of functions that can be run in parallel in the pipe"""
```
let's move this stuff to a new PR as you suggested, I have marked the test as skipped and we can discuss there.

sh-rp commented on dlt/extract/pipe.py:
```diff
@@ -773,91 +792,63 @@ def _resolve_futures(self) -> ResolvablePipeItem:
             # we need to decrease the index to keep the round robin order
-            self._round_robin_index -= 1
-            # since in this case we have popped an initial source, we need to decrease the initial sources count
-            self._initial_sources_count -= 1
```
this is actually a bug i would say..

sh-rp commented on tests/extract/test_extract_pipe.py:
```diff
@@ -603,6 +603,10 @@ def pass_gen(item, meta):
 def test_close_on_async_exception() -> None:
+    global close_pipe_got_exit, close_pipe_yielding
+    close_pipe_got_exit = False
+    close_pipe_yielding = False
```
globals need to be reset in each test, they will not recover between tests in one file!

sh-rp commented on tests/extract/test_extract_pipe.py:
```diff
@@ -674,6 +699,33 @@ def raise_gen(item: int):
+                yield i
+            close_pipe_yielding = False
+        # we have a different exception here
+        except asyncio.CancelledError:
```
i expected to also get a generatorexit here, but it seems to not be the case.

rudolfix replied:
@z3z1ma right! but we wrap async generator in regular generator so I expect GeneratorExit here. if we get both it means that we are leaking some edge cases ie. when pool is closed and we cancel async generators using asyncio module. I'll investigate in separate branch

sh-rp commented on tests/extract/test_extract_pipe.py:
```diff
@@ -715,13 +717,14 @@ async def long_gen():
             close_pipe_yielding = False
         # we have a different exception here
-        except asyncio.CancelledError:
+        except GeneratorExit:
```
I don't think this works. the test fails now. The error on cancelling async gens is a cancellederror, not a generatorexit.

rudolfix replied:
huh this was working for me. I was not cancelling async gen directly but the wrapped gen and propagating the exception. I'll check it out

sultaniman commented on dlt/extract/pipe.py:
```diff
@@ -498,20 +504,20 @@ def __init__(
+        self._sources = sources
         self._futures: List[FuturePipeItem] = []
-        self._next_item_mode = next_item_mode
+        self._next_item_mode: TPipeNextItemMode = next_item_mode
```
since it is annotated in the argument should we also remove the annotation here?

rudolfix replied:
good catch! but somehow mypy sees this as string here so I forced the type (or maybe it is VSCode language server)

sultaniman commented on dlt/extract/pipe.py:
```diff
@@ -773,91 +793,70 @@ def _resolve_futures(self) -> ResolvablePipeItem:
+                # get next item from the current source
+                gen, step, pipe, meta = self._sources[self._current_source_index]
+                set_current_pipe_name(pipe.name)
+                if (item := next(gen)) is not None:
```
imo this already will check for not None as `item := next(gen)`

sultaniman commented on dlt/extract/pipe.py:
```diff
@@ -773,91 +793,70 @@ def _resolve_futures(self) -> ResolvablePipeItem:
+                if first_evaluated_index is None:
+                    first_evaluated_index = self._current_source_index
+                # always go round robin if None was returned
+                self._current_source_index = (self._current_source_index - 1) % sources_count
```
should we also check if `self._current_source_index > 0`?

rudolfix replied:
no! check how the modulo operator works for negative numbers :)

sultaniman commented on dlt/extract/utils.py:
```diff
@@ -119,6 +132,45 @@ def check_compat_transformer(name: str, f: AnyFun, sig: inspect.Signature) -> in
+) -> Generator[Awaitable[TDataItems], None, None]:
+    """Wraps an async generator into a list of awaitables"""
+    exhausted = False
+    busy = False
```
is it better to pass some object instance to `run` function which will contain the state the variables?
```py
    exhausted = False
    busy = False
```

rudolfix commented on dlt/extract/utils.py:
```diff
@@ -119,6 +132,45 @@ def check_compat_transformer(name: str, f: AnyFun, sig: inspect.Signature) -> in
+            nonlocal busy
+            busy = False
+
+    # this generator yields None while the async generator is not exhausted
```
in a sense. most of this PR was to handle exactly this case. it is done upstream, if we see that all generators want to idle and we have no more data we sleep and yield control to other threads

rudolfix reviewed (approved):
LGTM! thanks @sh-rp and everyone else that participated and gave us ideas to implement that!
